src/mbgl/text/glyph_set.cpp

```diff
@@ -87,83 +86,113 @@ void justifyLine(std::vector<PositionedGlyph> &positionedGlyphs, const std::map<
-
-                uint32_t lineLength = positionedGlyphs[lastSafeBreak + 1].x;
-                maxLineLength = util::max(lineLength, maxLineLength);
-
-                for (uint32_t k = lastSafeBreak + 1; k <= i; k++) {
-                    positionedGlyphs[k].y += lineFeedOffset;
-                    positionedGlyphs[k].x -= lineLength;
-                }
-
-                if (justify) {
-                    // Collapse invisible characters.
-                    uint32_t breakGlyph = positionedGlyphs[lastSafeBreak].glyph;
-                    uint32_t lineEnd = lastSafeBreak;
-                    if (util::i18n::isVisible(breakGlyph)) {
-                        lineEnd--;
-                    }
-
-                    justifyLine(positionedGlyphs, sdfs, lineStartIndex, lineEnd, justify);
-                }
-
-                lineStartIndex = lastSafeBreak + 1;
-                lastSafeBreak = 0;
-                lengthBeforeCurrentLine += lineLength;
-                line++;
-            }
-
-            // Ideographic characters, spaces, and word-breaking punctuation that often appear without surrounding spaces.
-            if (useBalancedIdeographicBreaking
-                || util::i18n::allowsWordBreaking(shape.glyph)
-                || util::i18n::allowsIdeographicBreaking(shape.glyph)) {
-                lastSafeBreak = i;
-            }
+        if ( currentX > maxWidth && lastSafeBreak > 0 )
+        {
+            lineBreakPoints.insert(lastSafeBreak);
+            currentX -= lastSafeBreakX;
+            lastSafeBreakX = 0;
         }
+        
+        currentX += glyph.metrics.advance + spacing;
     }
 
-    const PositionedGlyph& lastPositionedGlyph = positionedGlyphs.back();
-    const auto lastGlyphIt = sdfs.find(lastPositionedGlyph.glyph);
-    assert(lastGlyphIt != sdfs.end());
-    const uint32_t lastLineLength = lastPositionedGlyph.x + lastGlyphIt->second.metrics.advance;
-    maxLineLength = std::max(maxLineLength, lastLineLength);
-
-    const uint32_t height = (line + 1) * lineHeight;
-
-    justifyLine(positionedGlyphs, sdfs, lineStartIndex, uint32_t(positionedGlyphs.size()) - 1, justify);
-    align(shaping, justify, horizontalAlign, verticalAlign, maxLineLength, lineHeight, line, translate);
+    return lineBreakPoints;
+}
+    
+void GlyphSet::shapeLines(Shaping& shaping, const std::vector<std::u16string>& lines,
+                             const float spacing, const float lineHeight, const float horizontalAlign, const float verticalAlign,
+                             const float justify, const Point<float> &translate) const {
 
+    // the y offset *should* be part of the font metadata
+    const int32_t yOffset = -17;
+    
+    float x = 0;
+    float y = yOffset;
+    
+    float maxLineLength = 0;
+    
+    for (std::u16string line : lines)
+    {
+        // Collapse whitespace so it doesn't throw off justification
+        boost::algorithm::trim_if(line, boost::algorithm::is_any_of(u" \t\n\v\f\r"));
```

Comment:
This doesn’t include the ideographic space (U+3000) which is used in CJK, or some of the [more exotic Western spaces](https://en.wikipedia.org/wiki/Whitespace_character#Unicode) such as the thin space and mutton. Since we’re bringing in ICU, can we take advantage of any character class functionality in there? Or are we concerned about it being too slow?